plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -214,47 +214,101 @@ private void returnFailure(BulkByScrollResponse response, String modelId, Action
-        DeleteRequest deleteRequest = new DeleteRequest(ML_MODEL_INDEX, modelId);
-        client.delete(deleteRequest, new ActionListener<DeleteResponse>() {
+    private void deleteModel(String modelId, FunctionName functionName, ActionListener<DeleteResponse> actionListener) {
+        DeleteRequest deleteRequest = new DeleteRequest(ML_MODEL_INDEX, modelId).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+        client.delete(deleteRequest, new ActionListener<>() {
             @Override
             public void onResponse(DeleteResponse deleteResponse) {
-                deleteModelChunks(modelId, deleteResponse, actionListener);
-                deleteController(modelId);
+                deleteModelChunksAndController(functionName, actionListener, modelId, deleteResponse);
             }
 
             @Override
             public void onFailure(Exception e) {
-                log.error("Failed to delete model meta data for model: " + modelId, e);
                 if (e instanceof ResourceNotFoundException) {
-                    deleteModelChunks(modelId, null, actionListener);
-                    deleteController(modelId);
+                    deleteModelChunksAndController(functionName, actionListener, modelId, null);
+                } else {
+                    log.error("Failed to delete model meta data for model, please try again: " + modelId, e);
+                    actionListener.onFailure(e);
                 }
-                actionListener.onFailure(e);
             }
         });
     }
 
+    private void deleteModelChunksAndController(
+        FunctionName functionName,
+        ActionListener<DeleteResponse> actionListener,
+        String modelId,
+        DeleteResponse deleteResponse
+    ) {
+        if (FunctionName.REMOTE != functionName) {
+            CountDownLatch countDownLatch = new CountDownLatch(2);
+            AtomicBoolean bothDeleted = new AtomicBoolean(true);
+            ActionListener<Boolean> countDownActionListener = ActionListener.wrap(b -> {
+                countDownLatch.countDown();
+                bothDeleted.compareAndSet(true, b);
+                if (countDownLatch.getCount() == 0) {
+                    if (bothDeleted.get()) {
+                        log.debug("model chunks and model controller for model {} deleted successfully", modelId);
+                        if (deleteResponse != null) {
+                            actionListener.onResponse(deleteResponse);
+                        } else {
+                            actionListener.onFailure(new OpenSearchStatusException("Failed to find model", RestStatus.NOT_FOUND));
+                        }
+                    } else {
+                        actionListener
+                            .onFailure(
+                                new IllegalStateException("Failed to delete model chunks or model controller, please try again: " + modelId)
```

Comment:
This looks complicated for user to understand which one failed. 
How about just delete model chunk first, then delete model controller ? Throw exception if any delete fails

Reply:
How about refine this error message to more generic, e.g. "model is not all cleaned up, please try again"?

Reply:
Sounds good